session/tlv.go

```diff
@@ -152,3 +154,178 @@ func DeserializeSession(r io.Reader) (*Session, error) {
 
 	return session, nil
 }
+
+// encodeMacaroonRecipe serializes a macaroon recipe so that it can be stored
+// as a tlv value. The serialization will follow the following pattern:
+//
+//	| numPerms | perm1Len | perm1Bytes | perm2Len | perms2Bytes |...
+// 	| numCaveats | caveat1IDLen | caveat1ID | caveat1VerificationIDLen
+//	| caveat1VerificationID | caveat1LocationLen | caveat1Location | ... |
+func encodeMacaroonRecipe(w io.Writer, val interface{},
+	buf *[8]byte) error {
+
+	if v, ok := val.(*MacaroonRecipe); ok {
+		// Encode the macaroon permissions. We do this by first writing
+		// the number of permissions that will be encoded.
+		numPerms := uint64(len(v.Permissions))
+		if err := tlv.WriteVarInt(w, numPerms, buf); err != nil {
+			return err
+		}
+
+		// Now, we iterate over the permissions and for each, we convert
+		// the permission to a string of form `<entity>:<action>` which
+		// we then convert to a byte slice. We first write the length
+		// of the byte slice, followed by the byte slice.
+		for _, perm := range v.Permissions {
+			permBytes := []byte(permissionToString(perm))
```

Comment:
Nice simplification! I guess we could even have concatenated all permission pairs together, separated by a comma. But this is also good and adds one less restriction on the name/content of the entity/action pair.